jobs: child proc must have a separate process-group

justinmk opened the pull request:
closes #6530 

- are PTY jobs being killed correctly?
- ~~is #6891 related?~~
- [x] win: iterate through children (`os_proc_tree_kill()`)
- [x] implement new API function: `nvim_get_proc_children()`
- [x] implement new API function: `nvim_get_proc()`
- [x] tests

Didn't fix this common test failure (maybe #7376 ?):

```
TermClose event reports the correct <abuf>: -- Output to stderr:
Vim: Caught deadly signal 'SIGHUP'
Vim: Finished.
CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):
  Running functional tests failed with error: 1.
...
===============================================================================
NVIM_LOG_FILE: /home/travis/build/neovim/neovim/build/.nvimlog
2018/03/06 08:57:28 ERROR 11479 loop_close:133: uv_loop_close() hang?
[--I] signal   0xa310e8
[-AI] async    0xa30f30
[R--] signal   0xa2a620
```

quickbuild failure is unrelated, `spell_spec.lua` #8102

justinmk committed:
jobs: child proc must have a separate process-group

UV_PROCESS_DETACHED compels libuv:uv__process_child_init() to call
setsid() in the child just after fork().  That ensures the process and
its descendants are grouped in a separate session (and process group).

The following jobstart() call correctly groups `sh` and `sleep` in a new
session (and process-group), where `sh` is the "session leader" (and
process-group leader):

    :call jobstart(['sh','-c','sleep 60'])

     SESN  PGRP   PID  PPID  Command
    30383 30383 30383  3620  │  ├─ -bash
    30383 31432 31432 30383  │  │  └─ nvim -u NORC
    30383 31432 31433 30383  │  │     ├─ nvim -u NORC
     8105  8105  8105 31432  │  │     └─ sh -c sleep 60
     8105  8105  8106  8105  │  │        └─ sleep 60

closes #6530
ref: https://stackoverflow.com/q/1046933
ref: https://unix.stackexchange.com/a/404065

Helped-by: Marco Hinz <mh.codebro+github@gmail.com>

Discussion
------------------------------------------------------------------------

On my linux box before this patch, the termclose_spec.lua:'kills job
trapping SIGTERM' test indirectly causes cmake/busted to wait for 60s.
That's because the test spawns a `sleep 60` descendant process which
hangs around even after nvim exits: nvim killed the parent PID, but not
PGID (process-group), so the grandchild "reparented" to init (PID 1).

Session contains processes (and process-groups) which are logically part
of the same "login session". Process-group is a set of
logically/informally-related processes within a session; for example,
shells assign a process group to each "job". Session IDs and PGIDs both
have type pid_t (like PIDs).

These OS-level mechanisms are, as usual, legacy accidents whose purpose
is upheld by convention and folklore.  We can use session-level grouping
(setsid), or we could use process-group-level grouping (setpgid).

Vim uses setsid() if available, otherwise setpgid(0,0).

Windows
------------------------------------------------------------------------

UV_PROCESS_DETACHED on win32 sets CREATE_NEW_PROCESS_GROUP flag.
But uv_kill() does not kill the process-group:
https://github.com/nodejs/node/issues/3617

Ideas:
- Set UV_PROCESS_DETACHED (CREATE_NEW_PROCESS_GROUP), then call
  GenerateConsoleCtrlEvent(CTRL_BREAK_EVENT, pid)
   - Maybe won't work because MSDN says "Only processes that share the
     same console as the calling process receive the signal."
     https://docs.microsoft.com/en-us/windows/console/generateconsolectrlevent
     But CREATE_NEW_PROCESS_GROUP creates a new console ...
     ref https://stackoverflow.com/q/1453520
- Group processes within a "job". libuv does that *globally* for
  non-detached processes: uv__init_global_job_handle.
- Iterate through CreateToolhelp32Snapshot().
   - https://stackoverflow.com/q/1173342
   - Vim does this, see terminate_all()

justinmk committed:
test: use luv.now() instead of os.time()

mhinz commented:
```vim
:let id = jobstart('sleep 30 | sleep 30 | sleep 30')
:call jobstop(id)
```

..finally works as expected after this patch. :+1:

oni-link commented on src/nvim/os/pty_process_unix.c:
```diff
@@ -145,8 +145,12 @@ void pty_process_teardown(Loop *loop)
+  FUNC_ATTR_NONNULL_ALL  // FUNC_ATTR_NORETURN ?
 {
+  // New session and progress-group. #6530
+  setsid();
```
Why a new session instead of a new process group? Would `setpgid(0,0);` also do the job?

bfredl replied:
Isn't a new session customary in a terminal emulator?

oni-link commented on src/nvim/event/process.c:
```diff
@@ -249,15 +250,15 @@ static void children_kill_cb(uv_timer_t *handle)
-      ILOG("Sending %s to pid %d", sig == SIGTERM ? "SIGTERM" : "SIGKILL",
-           proc->pid);
-      uv_kill(proc->pid, sig);
+      ILOG("sending %s to pid %d", sig == SIGTERM ? "SIGTERM" : "SIGKILL", pid);
```
pid -> pgid

justinmk committed:
win: os_proc_tree_kill()

XXX: comment at https://stackoverflow.com/q/1173342 :
> Windows recycles PIDs quite fast, you have to be extra careful not
> to kill unrelated processes. These APIs will report PPIDs for long
> dead processes whose PIDs may have been recycled. Check the parent
> start date to make sure it is related to the processes you spawned.

justinmk committed:
API: nvim_get_proc_children()

ref https://github.com/libuv/libuv/pull/836

janlazo commented:
Will this fix the `invalid channel id` or the exit code for job tests for cat.exe on Windows?

justinmk commented:
> Will this fix the invalid channel id or the exit code for job tests for cat.exe on Windows?

I wouldn't expect this to fix "invalid channel id". And, I don't think it will change behavior of cat.exe unless cat.exe was running in a shell.

> I'm using jobclose to get exit code 0 because jobstop can cause exit code 1.

We should probably keep that test "pending" then :)

jamessan commented on src/nvim/os/process.c:
```diff
@@ -0,0 +1,168 @@
+  bool exists = false;
+  size_t p_count = len / sizeof(*p_list);
+  for (size_t i = 0; i < p_count; i++) {
+    exists |= (p_list[i].kp_proc.p_pid == ppid);
```
Shouldn't this be

```c
exists ||= (p_list[i].ki_pid == ppid);
if (p_list[i].ki_ppid == ppid) {
  temp = xrealloc(temp, (*proc_count + 1) * sizeof(*temp));
  temp[*proc_count] = p_list[i].ki_pid;
```

oni-link commented on src/nvim/os/process.c:
```diff
@@ -0,0 +1,178 @@
+  snprintf(proc_p, sizeof(proc_p), "/proc/%d/task/%d/children", ppid, ppid);
+  FILE *fp = fopen(proc_p, "r");
+  if (fp == NULL) {
+    return 1;  // Process not found.
```
I'm missing the `children` entry under `proc`, because of a unset kernel option. Could one fallback to `pgrep -P ppid` or `ps --ppid ppid` in this case?

justinmk replied:
@oni-link How does `ps` get this info without the kernel providing an API?

oni-link replied:
Perhaps iterating through the entries (processes) under `/proc` and evaluating each entry `status` for `ppid`?

justinmk replied:
@oni-link Added a fallback for `pgrep` since it seems to be available on linux and macOS, and hopefully xBSD.

justinmk committed:
nvim_get_proc_children: fallback to shell

/proc/…/children may be unavailable because of an unset kernel option.
Fallback to `pgrep` invoked in a shell.

justinmk committed:
win: nvim_get_proc_children()

TODO: Raymond Chen explains[1] racy behavior of the
CreateToolhelp32Snapshot approach.  Better approach:

> create a job object and put process P in it. Then call
> QueryInformationJobObject with JobObjectBasicProcessIdList to get the
> list of child processes.

[1] "Why is CreateToolhelp32Snapshot returning incorrect parent process IDs all of a sudden?"
    https://blogs.msdn.microsoft.com/oldnewthing/20150403-00/?p=44313

justinmk committed:
API: nvim_get_proc()

TODO: "exepath" field (win32: QueryFullProcessImageName())

On unix-likes `ps` is used because the platform-specific APIs are
a nightmare.  For reference, below is a (incomplete) attempt:

diff --git a/src/nvim/os/process.c b/src/nvim/os/process.c
index 09769925aca5..99afbbf290c1 100644
--- a/src/nvim/os/process.c
+++ b/src/nvim/os/process.c
@@ -208,3 +210,60 @@ int os_proc_children(int ppid, int **proc_list, size_t *proc_count)
   return 0;
 }

+/// Gets various properties of the process identified by `pid`.
+///
+/// @param pid Process to inspect.
+/// @return Map of process properties, empty on error.
+Dictionary os_proc_info(int pid)
+{
+  Dictionary pinfo = ARRAY_DICT_INIT;
+#ifdef WIN32
+
+#elif defined(__APPLE__)
+  char buf[PROC_PIDPATHINFO_MAXSIZE];
+  if (proc_pidpath(pid, buf, sizeof(buf))) {
+    name = getName(buf);
+    PUT(pinfo, "exepath", STRING_OBJ(cstr_to_string(buf)));
+    return name;
+  } else {
+    ILOG("proc_pidpath() failed for pid: %d", pid);
+  }
+#elif defined(BSD)
+# if defined(__FreeBSD__)
+#  define KP_COMM(o) o.ki_comm
+# else
+#  define KP_COMM(o) o.p_comm
+# endif
+  struct kinfo_proc *proc = kinfo_getproc(pid);
+  if (proc) {
+    PUT(pinfo, "name", cstr_to_string(KP_COMM(proc)));
+    xfree(proc);
+  } else {
+    ILOG("kinfo_getproc() failed for pid: %d", pid);
+  }
+
+#elif defined(__linux__)
+  char fname[256] = { 0 };
+  char buf[MAXPATHL];
+  snprintf(fname, sizeof(fname), "/proc/%d/comm", pid);
+  FILE *fp = fopen(fname, "r");
+  // FileDescriptor *f = file_open_new(&error, fname, kFileReadOnly, 0);
+  // ptrdiff_t file_read(FileDescriptor *const fp, char *const ret_buf,
+  //                     const size_t size)
+  if (fp == NULL) {
+    ILOG("fopen() of /proc/%d/comm failed", pid);
+  } else {
+    size_t n = fread(buf, sizeof(char), sizeof(buf) - 1, fp);
+    if (n == 0) {
+      WLOG("fread() of /proc/%d/comm failed", pid);
+    } else {
+      size_t end = MIN(sizeof(buf) - 1, n);
+      end = (end > 0 && buf[end - 1] == '\n') ? end - 1 : end;
+      buf[end] = '\0';
+      PUT(pinfo, "name", STRING_OBJ(cstr_to_string(buf)));
+    }
+  }
+  fclose(fp);
+#endif
+  return pinfo;
+}

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -1,3 +1,46 @@
+  return out
+end
+
+local function systemlist(...)
```
@neovim/core I'm considering documenting these in `:help lua` as part of the `vim` module. The normal lua idiom would be to return `status, rv` pair, where `status=false` on failure. But I think it's really useful to show the failed command with `error()`. Thoughts?

bfredl commented on src/nvim/lua/vim.lua:
```diff
@@ -1,3 +1,65 @@
+  return out
+end
+
+local function systemlist(cmd)
```
Do we really want to commit a new system interface that mixes `stdout` and `stderr` togheter per default?

justinmk replied:
@bfredl Thanks. I'll keep this internal for now.

bfredl commented on src/nvim/lua/vim.lua:
```diff
@@ -1,3 +1,65 @@
+local function system(cmd)
```
Until we have lua native job control this should probably use `systemlist` (or even better viml job control to distinguish stderr) internally and do a safe conversion to lua binary string, so that this function handles binary data correctly (unlike vimL `system()` where this is impossible)

teto replied:
or even remove the function to force using systemlist which is better in every way.

bfredl replied:
@teto this is not really true in Lua, the issue is just that vimL system() should not be used to define Lua system()

justinmk commented:
## todo / future reference

Discussion of `vim.system[list]()` in public Lua module:

- https://github.com/neovim/neovim/pull/8107#discussion_r175254868
- https://github.com/neovim/neovim/pull/8107#discussion_r175254916

janlazo commented:
Which `ps` is os_proc_info using on Windows?

justinmk commented:
> Which ps is os_proc_info using on Windows?

On WIN32 it uses the WIN32 API. No ps, no tasklist.

justinmk committed:
test: jobstop() kills entire process tree

Test correctly fails before 8d90171f8be6.
ref #6530

justinmk committed:
test/win: use cmd.exe for test

Can revert this after #8120.

janlazo commented:
@justinmk `ping -n 1 -w 30000` exits within a second.

justinmk commented:
@janlazo Gah, I changed that at the last minute to match https://github.com/neovim/neovim/blob/3bf57c117120327930fb39935d1672558f58d148/test/functional/terminal/ex_terminal_spec.lua#L51

I will fix after #8120 , when https://github.com/neovim/neovim/pull/8107/commits/ae409b5042abdbec67305a063cf921ddffcce0c8 can be reverted.

janlazo commented:
My bad on that commit. I used ping for <= 1 sec. timeout but any program that exits quickly would have sufficed because of how for loops work on cmd.exe. Powershell has a long startup (2-3 sec. to run `exit 4` for `jobwait()` test) so I couldn't use `Start-Sleep -Milliseconds 100`.
